crates/pipeline_manager/src/db/test.rs

```diff
@@ -0,0 +1,1140 @@
+use super::PipelineDescr;
+use super::{
+    storage::Storage, AttachedConnector, ConfigDescr, ConfigId, ConnectorDescr, ConnectorId,
+    ConnectorType, PipelineId, ProjectDB, ProjectDescr, ProjectId, ProjectStatus, Version,
+};
+use crate::db::DBError;
+use anyhow::Result as AnyResult;
+use async_trait::async_trait;
+use chrono::DateTime;
+use pretty_assertions::assert_eq;
+use proptest::prelude::*;
+use proptest::test_runner::{Config, TestRunner};
+use proptest_derive::Arbitrary;
+use std::collections::BTreeMap;
+use std::fmt::Debug;
+use std::sync::atomic::{AtomicU16, Ordering};
+use std::time::SystemTime;
+use tempfile::TempDir;
+use tokio::sync::Mutex;
+
+static DB_PORT_COUNTER: AtomicU16 = AtomicU16::new(5555);
+
+struct DbHandle {
+    db: ProjectDB,
+    _temp_dir: TempDir,
+}
+
+impl Drop for DbHandle {
+    fn drop(&mut self) {
+        // We drop `pg` before the temp dir gets deleted (which will
+        // shutdown postgres). Otherwise postgres log an error that the
+        // directory is already gone during shutdown which could be
+        // confusing for a developer.
+        if let Some(pg) = self.db.pg.as_mut() {
+            let _r = async {
+                pg.stop_db().await.unwrap();
+            };
+        }
+    }
+}
+
+async fn test_setup() -> DbHandle {
+    let _temp_dir = tempfile::tempdir().unwrap();
+    let temp_path = _temp_dir.path();
+
+    let port = DB_PORT_COUNTER.fetch_add(1, Ordering::Relaxed);
```

Comment:
This is fragile.  You don't want to assume that all these ports are available.  Having said that there is no easy way to locate unused ports.  This is the best I could do:

```rust
use std::net::TcpListener;

let listener = TcpListener::bind(("127.0.0.1", 0))?;
let port = listener.local_addr()?.port();
```
Problem is, once you drop the listener, the port can get stolen by another process.  Ideally Postgres itself would support dynamic port allocation in some form.

Reply:
I wouldn't spend too much time trying to fix this. Binding to an ephemeral port is an anti-pattern for long running servers anyway, and is something we'll have to fix later. I suggest the dev-mode form factor have a runner service on a fixed port that brings up pipelines locally.

Reply:
Fixed